aten/src/ATen/native/Convolution.cpp

```diff
@@ -343,55 +368,52 @@ at::Tensor convolution(
                           ctx.benchmarkCuDNN(), ctx.deterministicCuDNN(), ctx.userEnabledCuDNN());
 }
 
-static inline std::vector<int64_t> convolution_expand_param_if_needed(
-  IntArrayRef list_param, const char *param_name, int64_t expected_dim) {
-  if (list_param.size() == 1) {
-    return std::vector<int64_t>(expected_dim, list_param[0]);
-  } else if ((int64_t) list_param.size() != expected_dim) {
-    std::ostringstream ss;
-    ss << "expected " << param_name << " to be a single integer value or a "
-       << "list of " << expected_dim << " values to match the convolution "
-       << "dimensions, but got " << param_name << "=" << list_param;
-    AT_ERROR(ss.str());
-  } else {
-    return list_param.vec();
-  }
-}
-
 at::Tensor _convolution(
     const Tensor& input_r, const Tensor& weight_r, const Tensor& bias_r,
     IntArrayRef stride_, IntArrayRef padding_, IntArrayRef dilation_,
     bool transposed_, IntArrayRef output_padding_, int64_t groups_,
     bool benchmark, bool deterministic, bool cudnn_enabled) {
 
-  auto input = input_r.contiguous();
+  const bool input_is_mkldnn = (input_r.type_id() == MkldnnCPUTensorId());
+  auto input = input_r;
+  if (!input_is_mkldnn) {
+    input = input.contiguous();
+  }
   auto weight = weight_r;
   auto bias = bias_r;
   auto k = weight.ndimension();
+  // mkldnn conv2d weights could have been re-ordered to 5d by
+  // mkldnn_reorder_conv2d_weight
+  if (input_is_mkldnn && (k == input.ndimension() + 1)) {
```

Comment:
can't we just take input.ndimension in all cases?

Reply:
@dzhulgakov nope because for normal cpu and gpu paths, weight ndims is expected to be the same as input ndims.